akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2151,8 +2151,12 @@ private[akka] object TakeWithin {
       override def onPull(): Unit = pull(in)
       def onFailure(ex: Throwable): Unit =
         if ((maximumRetries < 0 || attempt < maximumRetries) && pf.isDefinedAt(ex)) {
-          switchTo(pf(ex))
-          attempt += 1
+          pf(ex) match {
+            case source if source eq Source.empty => completeStage()
+            case other =>
+              switchTo(other)
+              attempt += 1
+          }
```

Comment:
Looks like an unrelated change?

Reply:
When the fallback source is empty then avoid the materialization?

Reply:
Sorry, GitHub PR view tricked me into thinking this was some other operator. Yeah, that makes sense.

Reply:
As this PR may need more time, I will extract this change to a separate pr.